Fix scaled_dot_product_attention reference implementation to match MATH backend

tom-pollak opened the pull request:
- Fixes #119188 (exact match on MATH backend)
- Fixes #123911 (document float32 upcasting behaviour)

### Summary

The documented [reference implementation](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) of SDPA doesn't numerically match the MATH. This causes confusion when testing numerical accuracy of kernels / code vs MATH.

### Changes

Updated the reference implementation to match MATH's actual behavior. The key corrections are:

- The MATH backend pre-scales both query and key tensors before matmul for numerical stability, rather than scaling after the matmul operation.
- The MATH backend internally upcasts to float32 for fp16/bf16 inputs, then converts back to original dtype at the end.

Added regression test `test_reference_implementation_bitwise_match_math_backend`. Test verifies exact bitwise match (rtol=0, atol=0) between the reference implementation and MATH.

tom-pollak committed:
Add SDPA regression test

tom-pollak committed:
Fix SDPA reference implementation

MATH backend scales Q K pre-softmax, leading to numerical differences
when comparing the ref impl with MATH.

Now matches SDPBackend.MATH with `rotl=0., atol=0.`

tom-pollak commented:
@pytorchbot label "module: sdpa"

tom-pollak committed:
Fix lints

Just whitespace, but no longer a copy-paste from docs

jbschlosser reviewed (commented):
Thanks for the contribution! I agree with the fixes and it's nice to have bitwise-accurate validation against the math backend via tests.

One concern I have is that the test won't catch updates to the reference implementation in the docs. Running this via the doctest mechanism would address this. cc @svekars for insight on how to ensure this validation happens during doctest time

tom-pollak committed:
Remove dropout_p

Don't think they are seeded the same way on some archs, leading to
different dropout

tom-pollak committed:
Remove dropout_p (squash)

tom-pollak committed:
put upcast behind reduced precision flag

tom-pollak committed:
Add reduced precision flag param test

tom-pollak commented on torch/nn/functional.py:
```diff
@@ -5818,6 +5818,9 @@ def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.
                 is_causal=False, scale=None, enable_gqa=False) -> torch.Tensor:
             L, S = query.size(-2), key.size(-2)
             scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
+            origin_dtype = query.dtype
```
Added flag to doc and test, does this look ok?

tom-pollak committed:
clean up spacing

tom-pollak commented:
@drisspg Ok to be merged?

drisspg reviewed (approved):
Looks good, thanks

tom-pollak commented:
@pytorchbot merge

tom-pollak committed:
set temp_mask device

tom-pollak commented:
@drisspg my bad, needed to put `temp_mask` on the `attn_bias` device. Fixed now!

tom-pollak commented:
@pytorchmergebot merge

tom-pollak commented:
@drisspg Sorry, do you mind approving again? I just added a `device=attn_bias.device`
